web_poet/testing/pytest.py

```diff
@@ -34,38 +47,109 @@ def collect(self) -> Iterable[Union[pytest.Item, pytest.Collector]]:  # noqa: D1
     """Represents a directory containing one test."""
 
-    if _new_pytest:
-
-        def __init__(
-            self,
-            name,
-            parent=None,
-            config: Optional["pytest.Config"] = None,
-            session: Optional[pytest.Session] = None,
-            nodeid: Optional[str] = None,
-            **kw,
-        ) -> None:
-            super().__init__(name, parent, config, session, nodeid, **kw)
-            self.fixture = Fixture(self.path)
-
-    else:
-
-        def __init__(  # type: ignore[misc]
-            self,
-            name,
-            parent=None,
-            config: Optional[Any] = None,
-            session: Optional[pytest.Session] = None,
-            nodeid: Optional[str] = None,
-        ) -> None:
-            super().__init__(name, parent, config, session, nodeid)
-            self.fixture = Fixture(Path(self.fspath, self.name))
-
-    def runtest(self) -> None:  # noqa: D102
-        self.fixture.assert_output()
+    def __init__(self, name: str, parent=None, **kwargs) -> None:
+        super(WebPoetCollector, self).__init__(name, parent, **kwargs)
+        self.fixture = Fixture(self._path)
+
+    def collect(self) -> Iterable[Union[pytest.Item, pytest.Collector]]:
+        """Return a list of children (items and collectors) for this
+        collection node."""
+        if self.config.getoption("WEB_POET_TEST_PER_ITEM", default=False):
+            return [
+                WebPoetItem.from_parent(parent=self, name="item", fixture=self.fixture)
+            ]
+        else:
+            field_tests = [
+                WebPoetFieldItem.from_parent(
+                    parent=self, name=field, fixture=self.fixture, field_name=field
+                )
+                for field in self.fixture.get_expected_output_fields()
+            ]
+            no_extra_fields_tests = [
+                WebPoetNoExtraFieldsItem.from_parent(
+                    parent=self, name="NO_EXTRA_FIELDS", fixture=self.fixture
+                )
+            ]
+            return field_tests + no_extra_fields_tests
+
+
+class _WebPoetItem(pytest.Item, _PathCompatMixin):
+    def __init__(self, *, fixture: Fixture, **kwargs) -> None:
+        super().__init__(**kwargs)
+        self.fixture = fixture
+
+
+class WebPoetItem(_WebPoetItem):
+    def runtest(self) -> None:
+        self.fixture.assert_full_item_correct()
+
+    def reportinfo(self):
+        return self._path, 0, f"{self.fixture.short_name}"
+
+    def repr_failure(self, excinfo, style=None):
+        if isinstance(excinfo.value, ItemValueError):
+            got, expected = excinfo.value.args
+            return comparison_error_message(
+                config=self.config,
+                op="==",
+                expected=expected,
+                got=got,
+                prefix="The output doesn't match.",
+            )
+        else:
+            return super().repr_failure(excinfo, style)
```

Comment:
Lines like this are the only ones which are missing the coverage. I think that's fine, because this code doesn't run normally; it's a safety measure, to get better output in case something strange happens (e.g. a bug in our plugin).